GH-15046 - Include tag, repo, commitSHA, force parameters in matterbuild's cutplugin slashcommand response

adolsalamanca opened the pull request:
#### Summary
Create a small string with cmd flags to include them in the different responses of matterbuild’s cutplugin slash command.


#### Ticket
  https://github.com/mattermost/mattermost-server/issues/15046

adolsalamanca committed:
Include flags in responses as expected

adolsalamanca committed:
Update go.sum after go mod tidy

adolsalamanca commented:
@ali-farooq0 @hanzei @utkuufuk
We should consider a change to join this PR and https://github.com/mattermost/matterbuild/pull/46 together so we solve this prior to conflicts.

Thanks.
Best Regards,

utkuufuk commented:
@adolsalamanca Good point 💯 

Alternatively, I don't mind resolving conflicts on my PR if this one gets merged first.

c.c. @hanzei @ali-farooq0

alifarooq0 commented on server/server.go:
```diff
@@ -455,19 +455,20 @@ func cutPluginCommandF(w http.ResponseWriter, slashCommand *MMSlashCommand, tag,
 	}
 
-	msg := fmt.Sprintf("Tag %s created. Waiting for the artifacts to sign and publish.\nWill report back when the process completes.\nGrab :coffee: and a :doughnut: ", tag)
+	cmdFlags := fmt.Sprintf("tag %s, repo %s, commitSHA %s, force %v", tag, repo, commitSHA, force)
```
Lets wrap the flags in the backticks, easier for users to copy paste.

Something like
```
Flags: tag `%s`, repo `%s`, commitSHA `%s`, force `%v`.
```

Also lets include it at the top as opposed to in the sentence. As it'll be easier to incorporate the `username` text after.
e.g

```
Flags: tag `%s`, repo `%s`, commitSHA `%s`, force `%v`.
Tag created. Waiting for the artifacts to sign and publish.\nWill report back when the process completes.\nGrab :coffee: and a :doughnut: 
```

alifarooq0 commented on server/server.go:
```diff
@@ -455,19 +455,20 @@ func cutPluginCommandF(w http.ResponseWriter, slashCommand *MMSlashCommand, tag,
 	if err := createTag(ctx, client, Cfg.GithubOrg, repo, tag, commitSHA); errors.Is(err, ErrTagExists) {
 		if !force {
-			WriteErrorResponse(w, NewError(fmt.Errorf("Tag %s already exists, not generating any artifacts. Use --force to regenerate artifacts.", tag).Error(), nil))
+			WriteErrorResponse(w, NewError(fmt.Errorf("Tag already exists, details %s. Not generating any artifacts. Use --force to regenerate artifacts. ", cmdFlags).Error(), nil))
```
Initially I was thinking to include the flags here as well, but thinking more about it, since we'll include the flags in the main message, maybe it's not required here. So lets skip.

alifarooq0 commented on server/server.go:
```diff
@@ -455,19 +455,20 @@ func cutPluginCommandF(w http.ResponseWriter, slashCommand *MMSlashCommand, tag,
 			return nil
 		}
-		msg = fmt.Sprintf("Tag %s exists. Waiting for the artifacts to sign and publish.\nWill report back when the process completes.\nGrab :coffee: and a :doughnut: ", tag)
+		msg = fmt.Sprintf("Tag exists, flags %s. Waiting for the artifacts to sign and publish.\nWill report back when the process completes.\nGrab :coffee: and a :doughnut: ", cmdFlags)
```
makes sense to include it here though 👍 
```
Flags: tag `%s`, repo `%s`, commitSHA `%s`, force `%v`.
Tag exists. Waiting for the artifacts to sign and...
```

alifarooq0 commented on server/server.go:
```diff
@@ -455,19 +455,20 @@ func cutPluginCommandF(w http.ResponseWriter, slashCommand *MMSlashCommand, tag,
 	}
 
-	WriteEnrichedResponse(w, "Plugin Release Process", msg, "#0060aa", IN_CHANNEL)
+	WriteEnrichedResponse(w, fmt.Sprintf("Plugin Release Process, flags %s", cmdFlags), msg, "#0060aa", IN_CHANNEL)
```
```
Flags: ....
Plugin Release Process
```

alifarooq0 commented on server/server.go:
````diff
@@ -500,7 +501,7 @@ git push --set-upstream origin %[3]s
 `, tag, repo, branch)
 		url := fmt.Sprintf("https://github.com/mattermost/mattermost-marketplace/compare/production...%s?quick_pull=1&labels=3:+QA+Review,2:+Dev+Review", branch)
-		msg = fmt.Sprintf("Plugin was successfully signed and uploaded to Github and S3.\nTag: **%s**\nRepo: **%s**\n[Release Link](%s)\nTo add this release to the Plugin Marketplace run inside your local Marketplace repository:\n```%s\n```\nUse %s to open a Pull Request.", tag, repo, releaseURL, marketplaceCommand, url)
+		msg = fmt.Sprintf("Plugin was successfully signed and uploaded to Github and S3.\nTag: **%s**\nRepo: **%s**\nCommitSHA: **%s**\nForce: **%v**\n[Release Link](%s)\nTo add this release to the Plugin Marketplace run inside your local Marketplace repository:\n```%s\n```\nUse %s to open a Pull Request.", tag, repo, commitSHA, force, releaseURL, marketplaceCommand, url)
````
👍

alifarooq0 reviewed (changes requested):
Thanks for your contribution @adolsalamanca! 🎉  Just a one comment on where to put the flags.

alifarooq0 commented:
Regarding resolving the conflicts, whichever is ready to go should go in first and the conflicts can be resolved in the other PR, since the change is pretty straight forward.

adolsalamanca committed:
Apply suggested changes

adolsalamanca committed:
Remove duplicated %s

adolsalamanca commented:
@ali-farooq0 Applied requested changes. Thanks for reviewing!

alifarooq0 commented on server/server.go:
```diff
@@ -454,20 +454,21 @@ func cutPluginCommandF(w http.ResponseWriter, slashCommand *MMSlashCommand, tag,
 	}
 
-	WriteEnrichedResponse(w, "Plugin Release Process", msg, "#0060aa", IN_CHANNEL)
+	WriteEnrichedResponse(w, fmt.Sprintf("%s\nPlugin Release Process", cmdFlags), msg, "#0060aa", IN_CHANNEL)
```
I think this will double include the cmd flags, lets remove it from here

```suggestion
	WriteEnrichedResponse(w, "Plugin Release Process", msg, "#0060aa", IN_CHANNEL)
```

adolsalamanca replied:
Ohh, I missed it. Sorry for that :/

alifarooq0 commented on server/server.go:
```diff
@@ -454,20 +454,21 @@ func cutPluginCommandF(w http.ResponseWriter, slashCommand *MMSlashCommand, tag,
 	if err := createTag(ctx, client, Cfg.GithubOrg, repo, tag, commitSHA); errors.Is(err, ErrTagExists) {
 		if !force {
-			WriteErrorResponse(w, NewError(fmt.Errorf("Tag %s already exists, not generating any artifacts. Use --force to regenerate artifacts.", tag).Error(), nil))
+			WriteErrorResponse(w, NewError(fmt.Errorf("%s\nTag already exists. Not generating any artifacts. Use --force to regenerate artifacts. ", cmdFlags).Error(), nil))
```
Since the initial msg contains all the cmd flags maybe lets not add it here 

```suggestion
			WriteErrorResponse(w, NewError(fmt.Errorf("Tag already exists. Not generating any artifacts. Use --force to regenerate artifacts. ", cmdFlags).Error(), nil))
```

hanzei replied:
```suggestion
	WriteErrorResponse(w, NewError(fmt.Errorf("Tag %s already exists. Not generating any artifacts. Use --force to regenerate artifacts.", tag).Error(), nil))
```

alifarooq0 reviewed (changes requested):
Thanks for the quick response @adolsalamanca! Just one more change and we're good to go 👍

alifarooq0 commented:
also looks like the lint check is failing.
```
@@ -454,7 +454,7 @@
 		WriteErrorResponse(w, NewError(err.Error(), nil))
 		return nil
 	}
-	
+
 	cmdFlags := fmt.Sprintf("Flags: tag `%s`, repo `%s`, commitSHA `%s`, force `%v`.", tag, repo, commitSHA, force)
 	msg := fmt.Sprintf("%s
Tag created. Waiting for the artifacts to sign and publish.
Will report back when the process completes.
Grab :coffee: and a :doughnut: ", cmdFlags)
 	if err := createTag(ctx, client, Cfg.GithubOrg, repo, tag, commitSHA); errors.Is(err, ErrTagExists) {
gofmt failure
make: *** [Makefile:23: gofmt] Error 1

Exited with code exit status 2
```

adolsalamanca committed:
Update server/server.go

But if we remove it from this message there will be no flags shown. 
Am I missing an initial message that will show the tags?

Co-authored-by: Ali Farooq <ali.farooq0@pm.me>

adolsalamanca committed:
Apply some changes

alifarooq0 commented on server/server.go:
```diff
@@ -455,19 +455,20 @@ func cutPluginCommandF(w http.ResponseWriter, slashCommand *MMSlashCommand, tag,
 	}
 
-	WriteEnrichedResponse(w, "Plugin Release Process", msg, "#0060aa", IN_CHANNEL)
+	WriteEnrichedResponse(w, "Plugin Release Process, flags", msg, "#0060aa", IN_CHANNEL)
```
It's ok to commit the flags part here, otherwise the heading will look odd.

This is how it currently looks

![image](https://user-images.githubusercontent.com/25732808/87972063-4d5afa80-ca94-11ea-802b-1ed49681d331.png)

```suggestion
	WriteEnrichedResponse(w, "Plugin Release Process", msg, "#0060aa", IN_CHANNEL)
```

alifarooq0 committed:
Update server/server.go

alifarooq0 reviewed (approved):
LGTM. Thanks for your contribution and patience @adolsalamanca! 🎉

adolsalamanca commented:
It's a pleasure :)

Just one minor thing: In line 462, seems like the flags will not be shown as neither the msg nor the cmdFlags are printed there. 
Is that ok, or I'm missing something in other method?

Thanks you so much for your help @ali-farooq0  :)

alifarooq0 commented:
Oh dang it, you're right @adolsalamanca. Good catch. I saw that it was an error and didn't see that we don't even print the other msg. Makes sense, if you can go ahead and make that change.

```
Flags: ...
Tag already exists. Not...
```

adolsalamanca committed:
Update error response

alifarooq0 reviewed (approved):
LGTM. Thanks for making the change @adolsalamanca! 🎉

adolsalamanca commented:
> LGTM. Thanks for making the change @adolsalamanca! 🎉

Thank you too :)

hanzei commented on server/server.go:
```diff
@@ -455,13 +455,14 @@ func cutPluginCommandF(w http.ResponseWriter, slashCommand *MMSlashCommand, tag,
 
-	msg := fmt.Sprintf("Tag %s created. Waiting for the artifacts to sign and publish.\nWill report back when the process completes.\nGrab :coffee: and a :doughnut: ", tag)
+	cmdFlags := fmt.Sprintf("Flags: tag `%s`, repo `%s`, commitSHA `%s`, force `%v`.", tag, repo, commitSHA, force)
+	msg := fmt.Sprintf("%s\nTag created. Waiting for the artifacts to sign and publish.\nWill report back when the process completes.\nGrab :coffee: and a :doughnut: ", cmdFlags)
```
Instead of plain dumping the flags could the text be more verbal? Something like `Tag $TAG created in $REPO at commit $COMMIT_SHA.`

alifarooq0 replied:
See my comment below. https://github.com/mattermost/matterbuild/pull/45#discussion_r457697905. However 0/5 on the approach though. We can always iterate over this if need be. Dumping the flags was the easiest way to get them in :D

hanzei replied:
```suggestion
	username := slashCommand.Username
	msg := fmt.Sprintf("`@%s` triggered a plugin release process using `%s`.\nTag %s created in `%s`. Waiting for the artifacts to sign and publish.\nWill report back when the process completes.\nGrab :coffee: and a :doughnut: ", username, command.Text, repo, tag)
```

adolsalamanca replied:
@hanzei with command.Text you mean the text string that is inside slashCommand.Command?

hanzei replied:
`slashCommand.Command` does only contain `/mb` in this case. `command.Text` contains the text after `/mb`. Hence, you can get the full command by concatenating these two with a white space.

hanzei commented on server/server.go:
```diff
@@ -455,13 +455,14 @@ func cutPluginCommandF(w http.ResponseWriter, slashCommand *MMSlashCommand, tag,
 	}
 
-	msg := fmt.Sprintf("Tag %s created. Waiting for the artifacts to sign and publish.\nWill report back when the process completes.\nGrab :coffee: and a :doughnut: ", tag)
+	cmdFlags := fmt.Sprintf("Flags: tag `%s`, repo `%s`, commitSHA `%s`, force `%v`.", tag, repo, commitSHA, force)
```
@ali-farooq0 What values does dumping `force` here provide? Based on the response it should already be clear if it was passed.

alifarooq0 replied:
my main motivation for including all these flags was to quickly reconstruct the `/mb` command, which is why i sort of like all of them in one spot. But honestly 0/5 on the approach as long as they're included in the backticks.

hanzei commented on server/server.go:
````diff
@@ -500,7 +501,7 @@ git push --set-upstream origin %[3]s
 `, tag, repo, branch)
 		url := fmt.Sprintf("https://github.com/mattermost/mattermost-marketplace/compare/production...%s?quick_pull=1&labels=3:+QA+Review,2:+Dev+Review", branch)
-		msg = fmt.Sprintf("Plugin was successfully signed and uploaded to Github and S3.\nTag: **%s**\nRepo: **%s**\n[Release Link](%s)\nTo add this release to the Plugin Marketplace run inside your local Marketplace repository:\n```%s\n```\nUse %s to open a Pull Request.", tag, repo, releaseURL, marketplaceCommand, url)
+		msg = fmt.Sprintf("Plugin was successfully signed and uploaded to Github and S3.\nTag: **%s**\nRepo: **%s**\nCommitSHA: **%s**\nForce: **%v**\n[Release Link](%s)\nTo add this release to the Plugin Marketplace run inside your local Marketplace repository:\n```%s\n```\nUse %s to open a Pull Request.", tag, repo, commitSHA, force, releaseURL, marketplaceCommand, url)
````
Same question here: What does printing  `force` provide?

alifarooq0 replied:
I think it's helpful to include it in the final response to know how a build was initiated. But 0/5 if we should include the `force` flag or not.

hanzei replied:
I don't think it's valuable to have it here. The post above already makes clear how the release got created. @adolsalamanca Would you please remove it?

adolsalamanca commented:
Ok guys @hanzei @ali-farooq0 
I understand both points of view. Which approach do you prefer to follow?
I can easily modify it and we could move on with the PR.

Thanks,

hanzei commented on server/server.go:
```diff
@@ -455,13 +455,14 @@ func cutPluginCommandF(w http.ResponseWriter, slashCommand *MMSlashCommand, tag,
 			return nil
 		}
-		msg = fmt.Sprintf("Tag %s exists. Waiting for the artifacts to sign and publish.\nWill report back when the process completes.\nGrab :coffee: and a :doughnut: ", tag)
+		msg = fmt.Sprintf("%s\nTag exists. Waiting for the artifacts to sign and publish.\nWill report back when the process completes.\nGrab :coffee: and a :doughnut: ", cmdFlags)
```
```suggestion
		msg = fmt.Sprintf("Tag %s exists. Waiting for the artifacts to sign and publish.\nWill report back when the process completes.\nGrab :coffee: and a :doughnut: ", tag)
```

hanzei commented on server/server.go:
````diff
@@ -500,7 +501,7 @@ git push --set-upstream origin %[3]s
 `, tag, repo, branch)
 		url := fmt.Sprintf("https://github.com/mattermost/mattermost-marketplace/compare/production...%s?quick_pull=1&labels=3:+QA+Review,2:+Dev+Review", branch)
-		msg = fmt.Sprintf("Plugin was successfully signed and uploaded to Github and S3.\nTag: **%s**\nRepo: **%s**\n[Release Link](%s)\nTo add this release to the Plugin Marketplace run inside your local Marketplace repository:\n```%s\n```\nUse %s to open a Pull Request.", tag, repo, releaseURL, marketplaceCommand, url)
+		msg = fmt.Sprintf("Plugin was successfully signed and uploaded to Github and S3.\nTag: **%s**\nRepo: **%s**\nCommitSHA: **%s**\nForce: **%v**\n[Release Link](%s)\nTo add this release to the Plugin Marketplace run inside your local Marketplace repository:\n```%s\n```\nUse %s to open a Pull Request.", tag, repo, commitSHA, force, releaseURL, marketplaceCommand, url)
````
If no `commitSHA` is provided, can the commit sha of `master` get printed instead?

hanzei replied:
Printing the getting the sha of `master` is surely option. But could you omit the `CommitSHA:` if no `commitSHA` is provided.

hanzei reviewed (changes requested):
I and @ali-farooq0 synced up on this and got to a consensus. See my comments below.

I also included one change from https://github.com/mattermost/matterbuild/pull/46 to avoid a merge conflict.

adolsalamanca committed:
resolve conflict

adolsalamanca commented:
@hanzei @ali-farooq0 Applied suggested changes, the only thing I did not figure out was the commit SHA of master to get it printed if no commit SHA was provided.
Please feel free to add it or any other things.

Thanks.

hanzei reviewed (changes requested):
Could you please resolve the merge conflict and take a look at my comment?
- https://github.com/mattermost/matterbuild/pull/45#discussion_r462945059
- https://github.com/mattermost/matterbuild/pull/45#discussion_r462946303

looking forward to merging this PR soon!

adolsalamanca committed:
Apply suggested changes

adolsalamanca committed:
Fix renaming and correct typo

adolsalamanca committed:
Run formatting

adolsalamanca commented:
Sorry for the delay guys, I made a new branch from updated master and applied suggested changes.
Thanks

hanzei commented on server/server.go:
```diff
@@ -455,14 +455,13 @@ func cutPluginCommandF(w http.ResponseWriter, slashCommand *MMSlashCommand, tag,
 
-	username := slashCommand.Username
-	msg := fmt.Sprintf("@%s triggered a plugin release process.\nTag %s created. Waiting for the artifacts to sign and publish.\nWill report back when the process completes.\nGrab :coffee: and a :doughnut: ", username, tag)
+	msg := fmt.Sprintf("@%s triggered a plugin release process.\nTag %s created, Repo %s. Waiting for the artifacts to sign and publish.\nWill report back when the process completes.\nGrab :coffee: and a :doughnut: ", slashCommand.Username, tag, repo)
```
```suggestion
msg := fmt.Sprintf("@%s triggered a plugin release process using `%s`.\nTag %s created in  `%s`. Waiting for the artifacts to sign and publish.\nWill report back when the process completes.\nGrab :coffee: and a :doughnut: ", slashCommand.Username, command.Text, tag, repo)
```

hanzei commented on server/server.go:
```diff
@@ -455,14 +455,13 @@ func cutPluginCommandF(w http.ResponseWriter, slashCommand *MMSlashCommand, tag,
 	if err := createTag(ctx, client, Cfg.GithubOrg, repo, tag, commitSHA); errors.Is(err, ErrTagExists) {
 		if !force {
-			WriteErrorResponse(w, NewError(fmt.Errorf("@%s Tag %s already exists, not generating any artifacts. Use --force to regenerate artifacts.", username, tag).Error(), nil))
+			WriteErrorResponse(w, NewError(fmt.Errorf("@%s Tag %s already exists, repo %s, not generating any artifacts. Use --force to regenerate artifacts. ", slashCommand.Username, repo, tag).Error(), nil))
```
```suggestion
			WriteErrorResponse(w, NewError(fmt.Errorf("@%s Tag %s already exists in %s. Not generating any artifacts. Use --force to regenerate artifacts. ", slashCommand.Username, repo, tag).Error(), nil))
```

hanzei commented on server/server.go:
```diff
@@ -455,14 +455,13 @@ func cutPluginCommandF(w http.ResponseWriter, slashCommand *MMSlashCommand, tag,
 			return nil
 		}
-		msg = fmt.Sprintf("@%s Tag %s already exists. Waiting for the artifacts to sign and publish.\nWill report back when the process completes.\nGrab :coffee: and a :doughnut: ", username, tag)
+		msg = fmt.Sprintf("@%s Tag %s already exists, repo %s. Waiting for the artifacts to sign and publish.\nWill report back when the process completes.\nGrab :coffee: and a :doughnut: ", slashCommand.Username, tag, repo)
```
```suggestion
		msg = fmt.Sprintf("@%s Tag %s already exists in %s. Waiting for the artifacts to sign and publish.\nWill report back when the process completes.\nGrab :coffee: and a :doughnut: ", slashCommand.Username, tag, repo)
```

hanzei reviewed (changes requested):
Somehow my suggestion from https://github.com/mattermost/matterbuild/pull/45#discussion_r460887032 got lost. I've submitted them again.

alifarooq0 commented:
Hey @adolsalamanca, just checking to see if you're able to make the changes requested, really excited to get this change released to production :D

adolsalamanca committed:
Merge branch 'GH-15046' of github.com:adolsalamanca/matterbuild into GH-15046

adolsalamanca committed:
Apply suggested changes

adolsalamanca commented:
Sorry for the delay guys, I just applied suggested changes.
Thanks!

hanzei committed:
Merge remote-tracking branch 'origin/master' into GH-15046

hanzei reviewed (approved):
Thanks @adolsalamanca :+1:
